File: 113. Path Sum II/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    private void pathSumHelper(List<List<Integer>> result, List<Integer> current, TreeNode node, int sum) {
        if (node == null) {
            return;
        }

        current.add(node.val);
        if (node.left == null && node.right == null && node.val == sum) {
            result.add(new ArrayList<>(current));
        } else {
            pathSumHelper(result, current, node.left, sum - node.val);
            pathSumHelper(result, current, node.right, sum - node.val);
        }
        current.remove(current.size() - 1);
    }
    
    public List<List<Integer>> pathSum(TreeNode root, int sum) {
        List<List<Integer>> result = new ArrayList<>();
        List<Integer> current = new ArrayList<>();
        pathSumHelper(result, current, root, sum);
        return result;
    }
}

File: 113. Path Sum II/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def helper(self, node, s, ret, k):
        if not node:
            return
        ret.append(node.val)
        if not node.left and not node.right and node.val + s == k:
            self.rets.append(ret.copy())
        else:
            self.helper(node.left, s + node.val, ret, k)
            self.helper(node.right, s + node.val, ret, k)
        ret.pop()
            
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> List[List[int]]:
        self.rets = []
        self.helper(root, 0, [], targetSum)
        return self.rets
